### PR TITLE
Add a `filter_event_for_clients_with_state` function

### PR DESCRIPTION
**Pull request by erikjohnston**
This is going to be used in a refactor of the bulk push rule evaluator.

First we refactor existing methods to allow us to reuse the code.

Reviewable commit-by-commit.

### CODE REVIEW

File: synapse/visibility.py [hunk 1]
```diff
@@ -102,165 +105,369 @@ async def filter_events_for_client(
...
+    user_id: str,
+    event: EventBase,
+    visibility: str,
+    state: StateMap[EventBase],
+    is_peeking: bool,
+) -> bool:
+    """Check whether the user can see the event due to their membership
+
+    Returns:
+        True if they can, False if they can't
```
Comment (DMRobertson) [inline synapse/visibility.py:329]: There are now three `_check_*` functions with similar-but-different descriptions of their return values.

- _check_membership: True if user can definitely see the event, False if maybe not.
- _check_filter_to_send_to_client: True if might be allowed to be sent to clients, False if definitely not.
- _check_history_visibility: True if user can definitely see the event, False if maybe not.

This seems like a point of confusion to me. Could these functions' return values be modelled with an enum?

**Review by DMRobertson — COMMENTED** [review]
The "split out X" commits are a little hard to follow, I'm taking those on faith for the moment.

How thorough is our test coverage here?

**Comment by DMRobertson** [comment]
It also looks like there are a bunch of test failures :(

**Comment by erikjohnston** [comment]
Right! I think I hunted down the test failures. It was because we weren't correctly handling erased users.

**Comment by erikjohnston** [comment]
> The "split out X" commits are a little hard to follow, I'm taking those on faith for the moment.

Yup, sorry. They are literally just copy and paste of code.

File: synapse/visibility.py [hunk 2]
```diff
@@ -206,7 +208,7 @@ async def filter_event_for_clients_with_state(
     return {
         user_id
         for user_id in allowed_user_ids
-        if _check_membership(user_id, event, visibility, state_map, is_peeking)
+        if _check_membership(user_id, event, visibility, state_map, is_peeking)[0]
```
Comment (DMRobertson) [inline synapse/visibility.py]: Was this just always truthy before?

Comment (DMRobertson) [reply synapse/visibility.py]: (i.e. was this a bug waiting to happen?)

Comment (DMRobertson) [reply synapse/visibility.py]: Oh, ignore this. `_check_membership` was changed to return a tuple below.

File: synapse/visibility.py [hunk 3]
```diff
@@ -313,30 +322,39 @@ def _check_client_allowed_to_see_event(
...
     # If the sender has been erased and the user was not joined at the time, we
     # must only return the redacted form.
-    if sender_erased and membership != Membership.JOIN:
+    if sender_erased and not membership_result.joined:
         event = prune_event(event)
 
     return event
 
 
+@attr.s(frozen=True, auto_attribs=True)
```
Comment (DMRobertson) [inline synapse/visibility.py:344]: I think we usually do `slots=True` (though I doubt it saves us much memory here)

**Review by DMRobertson — APPROVED** [review]
Thanks, this looks sane. One small nit about `slots`.

Can you explain how this improves memory usage? Or perhaps the point is that this is prep work, and we'll call `filter_event_for_clients_with_state` in another PR?

**Comment by DMRobertson** [comment]
(thanks for [`9f5ed75` (#13222)](https://github.com/matrix-org/synapse/pull/13222/commits/9f5ed7546501d2c8341b41ec995d0e649e670eca) in particular)

**Comment by erikjohnston** [comment]
Oooh, that is exciting! Maybe we should upgrade our deps to use that

**Comment by DMRobertson** [comment]
> Oooh, that is exciting! Maybe we should upgrade our deps to use that

https://github.com/matrix-org/synapse/issues/13243

**Review by Shudogg — COMMENTED** [review]
Msg
